astropy/wcs/src/wcslib_celprm_wrap.c

```diff
@@ -0,0 +1,484 @@
+            self->x->flag = 0;
+        }
+    } else {
+        result = set_double("phi0", value, &phi0);
+        if (result) return result;
+        if (phi0 != self->x->phi0) {
+            self->x->phi0 = phi0;
+            self->x->flag = 0;
+        }
+    }
+    return 0;
+}
+
+
+static PyObject* PyCelprm_get_theta0(PyCelprm* self, void* closure)
+{
+    if (is_cel_null(self)) {
+        return NULL;
+    } else if (self->x->theta0 != UNDEFINED) {
+        return get_double("theta0", self->x->theta0);
+    }
+    Py_RETURN_NONE;
+}
+
+
+static int PyCelprm_set_theta0(PyCelprm* self, PyObject* value, void* closure)
+{
+    int result;
+    double theta0;
+    if(is_cel_null(self) || is_readonly(self)) {
+        return -1;
+    } else if (value == Py_None) {
+        if (self->x->theta0 != UNDEFINED) {
+            self->x->theta0 = UNDEFINED;
+            self->x->flag = 0;
+        }
+    } else {
+        result = set_double("theta0", value, &theta0);
+        if (result) return result;
+        if (theta0 != self->x->theta0) {
+            self->x->theta0 = theta0;
+            self->x->flag = 0;
+        }
+    }
+    return 0;
+}
+
+
+static PyObject* PyCelprm_get_ref(PyCelprm* self, void* closure)
+{
+    Py_ssize_t size = 4;
+    if (is_cel_null(self)) {
+        return NULL;
+    } else {
+        return get_double_array("ref", self->x->ref, 1, &size, (PyObject*) self);
+    }
+}
+
+
+static int PyCelprm_set_ref(PyCelprm* self, PyObject* value, void* closure)
+{
+    int i;
+    int skip[4] = {0, 0, 0, 0};
+    double ref[4] = {0.0, 0.0, UNDEFINED, +90.0};
+    npy_intp size;
+    double *data;
+
+    if (is_cel_null(self) || is_readonly(self)) return -1;
+
+    if (value == Py_None) {
+        /* If ref is set to None - reset ref to celini values: */
+        for (i = 0; i < 4; i++) {
+            self->x->ref[i] = ref[i];
+        }
+        self->x->flag = 0;
+        return 0;
+    }
+
+    PyObject* value_array = PyArray_ContiguousFromAny(value, NPY_DOUBLE, 1, 1);
+    if (!value_array) return -1;
+
+    size = PyArray_SIZE(value_array);
+
+    if (size < 1) {
+        Py_DECREF(value_array);
+        PyErr_SetString(PyExc_ValueError,
+            "'ref' must be a non-empty 1-dimentional list of values or None.");
+        return -1;
+    }
+
+    if (size > 4) {
+        Py_DECREF(value_array);
+        PyErr_SetString(PyExc_RuntimeError, "Number of 'ref' values cannot exceed 4.");
+        return -1;
+    }
+
+    if (PyList_Check(value)) {
+        for (i = 0; i < size; i++) {
+            skip[i] = (PyList_GetItem(value, i) == Py_None);
+        }
+    }
+
+    data = (double*) PyArray_DATA(value_array);
+
+    for (i = 0; i < size; i++) {
+        if (skip[i]) continue;
+        self->x->ref[i] = data[i];
+        if (npy_isnan(self->x->ref[i])) {
+            self->x->ref[i] = UNDEFINED;
+        } else {
+            self->x->ref[i] = data[i];
+        }
+    }
+    for (i = size; i < 4; i++) {
```

Comment:
Yes, it should. This is really for convenience. `ref` has 4 elements with the first two required ("given") and second two are "given and returned". Second two will be computed by the `set` routine if they are set to default values. This loop fills **missing values** (NOT ALL!!!) with defaults when user provides less than 4 values. For example, this allows doing `celprm.ref=wcs.crval` (or something like that) instead of `celprm.ref=[wcs.crval[0], wcs.crval[1], np.nan, 90]`.

Reply:
Ah ha!  Sorry, I missed the `i = size` instead of `i = 0`